src/services/factories/tokenList.ts

```diff
@@ -12,58 +12,106 @@ export function getTokensFactory(factoryParams: {
+      return { token, failed: true }
     }
   }
 
-  async function updateTokenIds(networkId: number): Promise<void> {
-    // Set as updated on start to prevent concurrent queries
-    updatedIdsForNetwork.add(networkId)
+  async function updateWithRetries(
+    networkId: number,
+    tokens: TokenDetails[],
+    maxRetries = 3,
+    retryDelay = 1000,
+    exponentialBackOff = true,
+  ): Promise<void> {
+    if (maxRetries <= 0) {
+      throw new Error(
+        `[network:${networkId}] Max retries exceeded trying to fetch token ids for tokens ${tokens.map(
+          t => t.address,
+        )}`,
+      )
+    }
 
-    const promises = tokenListApi.getTokens(networkId).map(token => getTokenId(networkId, token.address))
+    const promises = tokens.map(token => updateTokenId(networkId, token))
 
     const results = await Promise.all(promises)
 
-    let failedToUpdate = false
-
-    const tokenList = results.reduce((tokens: TokenDetails[], result: GetTokenIdResult, index: number) => {
-      if (result.value !== null) {
-        // We got a result, yay
-        const token = tokenListApi.getTokens(networkId)[index]
-        token.id = result.value
-        tokens.push(token)
-      } else if (!result.remove) {
-        // Failed to query, but don't remove from the list
-        tokens.push(tokenListApi.getTokens(networkId)[index])
-        // Try again next time
-        // TODO: try again only for the ones that failed
-        failedToUpdate = true
+    const updated = new Map<string, TokenDetails>()
+    const failedToUpdate: TokenDetails[] = []
+    const toRemove = new Set<string>()
+
+    // classify results into map/list/set for easier access
+    results.forEach(({ token, remove, failed }) => {
+      if (remove) {
+        toRemove.add(token.address)
+      } else if (failed) {
+        failedToUpdate.push(token)
+      } else {
+        updated.set(token.address, token)
       }
-      return tokens
-    }, [])
+    })
 
-    // If any token failed, clear flag to try again next time
-    if (failedToUpdate) {
-      updatedIdsForNetwork.delete(networkId)
-      // TODO: update only the ones that failed
+    // only bother update the list if there was anything updated/removed
+    if (toRemove.size > 0 || updated.size > 0) {
+      logDebug(`[network:${networkId}] Updated ${updated.size} ids and removed ${toRemove.size} tokens`)
+      // build a new list from current list of tokens
+      const tokenList = tokenListApi.getTokens(networkId).reduce<TokenDetails[]>((acc, token) => {
+        if (toRemove.has(token.address)) {
+          // removing tokens not registered in the exchange for this network
+          return acc
+        } else if (updated.has(token.address)) {
+          // updating
+          acc.push(updated.get(token.address) as TokenDetails)
+        } else {
+          // not changed OR failed to fetch
+          acc.push(token)
```

Comment:
```suggestion
        } else {
          // changed or old token
          acc.push(updated.get(token.address) || token)
```
not that it matters much